keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTSSampler.java

```diff
@@ -0,0 +1,651 @@
+            probAfterLeapfrog = ProbabilityCalculator.calculateLogProbFor(probabilisticVertices);
+            pThetaRAfterLeapFrog = probAfterLeapfrog - 0.5 * dotProduct(leapfrog.momentum);
+
+            logLikelihoodRatio = pThetaRAfterLeapFrog - pThetaR;
+        }
+
+        return stepsize;
+    }
+
+    private static double adaptStepSize(AutoTune autoTune, BuiltTree tree, int sampleNum) {
+
+        if (sampleNum <= autoTune.adaptCount) {
+
+            //1/(m+t0)
+            double percentageLeftToTune = (1 / (sampleNum + STABILISER));
+
+            //(1 - 1/(m+t0)) * Hm-1
+            double proportionalAcceptanceProb = (1 - percentageLeftToTune) * autoTune.averageAcceptanceProb;
+
+            //alpha/nu_alpha
+            double averageDeltaLikelihoodLeapfrog = tree.deltaLikelihoodOfLeapfrog / tree.treeSize;
+
+            //delta - alpha/nu_alpha
+            double acceptanceProb = autoTune.targetAcceptanceProb - averageDeltaLikelihoodLeapfrog;
+
+            //Hm = (1-1/(m+t0)) * Hm-1 + (1/(m+t0)) * (delta - (alpha/nu_alpha))
+            double updatedAverageAcceptanceProb = proportionalAcceptanceProb + (percentageLeftToTune * acceptanceProb);
+
+            //sqrt(mu)/gamma
+            double shrunkSampleCount = Math.sqrt(sampleNum) / SHRINKAGE_FACTOR;
+
+            //log(epsilon_m) = mu - (sqrt(m)/gamma) * Hm
+            double updatedLogStepSize = autoTune.shrinkageTarget - (shrunkSampleCount * updatedAverageAcceptanceProb);
+
+            //m^-k
+            double tendToZero = Math.pow(sampleNum, -TEND_TO_ZERO_EXPONENT);
+
+            //m^-k * log(epsilon_m)
+            double reducedStepSize = tendToZero * updatedLogStepSize;
+
+            //(1-m^-k) * log(epsilon_bar_m-1)
+            double increasedStepSizeFrozen = (1 - tendToZero) * autoTune.logStepSizeFrozen;
+
+            //log(epsilon_bar_m) = m^-k * log(epsilon_m) + (1 - m^-k) * log(epsilon_bar_m-1)
+            autoTune.logStepSizeFrozen = reducedStepSize + increasedStepSizeFrozen;
+
+            autoTune.averageAcceptanceProb = updatedAverageAcceptanceProb;
+            autoTune.logStepSize = updatedLogStepSize;
+
+            return Math.exp(autoTune.logStepSize);
+        } else {
+
+            return Math.exp(autoTune.logStepSizeFrozen);
+        }
+    }
+
+    static class LeapFrogged {
+        final Map<VertexId, DoubleTensor> position;
+        final Map<VertexId, DoubleTensor> momentum;
+        final Map<VertexId, DoubleTensor> gradient;
+
+        LeapFrogged(Map<VertexId, DoubleTensor> position,
+                    Map<VertexId, DoubleTensor> momentum,
+                    Map<VertexId, DoubleTensor> gradient) {
+            this.position = position;
+            this.momentum = momentum;
+            this.gradient = gradient;
+        }
+    }
+
+    static class BuiltTree {
+
+        Map<VertexId, DoubleTensor> positionBackward;
```

Comment:
Why don't these groups of maps use the `LeapFrogged` class? Makes it look much nicer having the `backward` and `forward` maps stored in one class.